### ISSUE TITLE
VMError$HostedError: Registering type as reachable after analysis

### ISSUE THREAD

**Issue by Twometer** [oracle/graal#4491]
**Describe the issue**
The build process crashes in the final stage

**Steps to reproduce the issue**
Please include both build steps as well as run steps
1. Create project with Spring Native and Spring Boot
2. Use the agent to generate reflection configs
3. Run `./gradlew nativeCompile`

**Describe GraalVM and your environment:**
 - GraalVM version [GraalVM CE 22.2.0-dev-20220413_2027](https://github.com/graalvm/graalvm-ce-dev-builds/releases/tag/22.2.0-dev-20220413_2027)
 - JDK major version: 11
 - OS: Amazon Linux 2
 - Architecture: aarch64

**More details**
```
[7/7] Creating image...                                                                                  (0.0s @ 7.37GB)
Fatal error: com.oracle.svm.core.util.VMError$HostedError: com.oracle.svm.core.util.VMError$HostedError: Registering type as reachable after analysis: AnalysisType<org.springframework.boot.web.embedded.tomcat.TomcatProtocolHandlerCustomizer[], allocated: false, inHeap: false, reachable: false>
        at com.oracle.svm.core.util.VMError.shouldNotReachHere(VMError.java:72)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:681)
        at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:518)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:407)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:585)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:615)
Caused by: com.oracle.svm.core.util.VMError$HostedError: Registering type as reachable after analysis: AnalysisType<org.springframework.boot.web.embedded.tomcat.TomcatProtocolHandlerCustomizer[], allocated: false, inHeap: false, reachable: false>
        at com.oracle.svm.core.util.VMError.shouldNotReachHere(VMError.java:68)
        at com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.registerAsReachable(AnalysisConstantReflectionProvider.java:332)
        at com.oracle.svm.hosted.ameta.HostedDynamicHubFeature.replace(HostedDynamicHubFeature.java:53)
        at com.oracle.graal.pointsto.meta.AnalysisUniverse.replaceObject(AnalysisUniverse.java:583)
------------------------------------------------------------------------------------------------------------------------
        at com.oracle.svm.hosted.image.NativeImageHeap.addArrayElements(NativeImageHeap.java:576)
        at com.oracle.svm.hosted.image.NativeImageHeap.addObjectToImageHeap(NativeImageHeap.java:473)
        at com.oracle.svm.hosted.image.NativeImageHeap.addObject(NativeImageHeap.java:295)
        at com.oracle.svm.hosted.image.NativeImageHeap.processAddObjectWorklist(NativeImageHeap.java:598)
        at com.oracle.svm.hosted.image.NativeImageHeap.addTrailingObjects(NativeImageHeap.java:198)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:663)
```

I also tried it under x86 Ubuntu with GraalVM version 22.0.0.2, which fails at the same stage, but with a NullPointerException. Without using the agent (letting spring generate the config), it works.

**Comment by oubidar-Abderrahim** [comment]
Hi, Thank you for reporting this we will take a look into it and get back to you

**Comment by Twometer** [comment]
I have an example repo that exhibits this problem here: https://github.com/Twometer/spring-native-data-bug-mve

**Comment by Twometer** [comment]
I've written a Python script that reads all `parameterTypes` and registers them in the `reflect-config` and now it compiles, so that probably is it.

**Comment by oubidar-Abderrahim** [comment]
Hi, could you please help me understand better the issue? 

does generating the reflect-config.json using the agent cause this issue while letting spring generates the config doesn't?

Could you please also share the commands to build, generate config in both cases and run the example? also sharing an example of the workaround you're using would be much appreciated

**Comment by Twometer** [comment]
Hi,

The workaround Python script we use to fix the configs is the following: https://gist.github.com/Twometer/b46a1f0713e0d4d724dc64630da2cdf6

Generating the reflect-config.json using the agent causes this issue, but the Spring Native generator does not exhibit this issue.

The command we use for creating the configs:
```
$JAVA_HOME/bin/java -DspringAot=true -agentlib:native-image-agent=config-merge-dir=src/main/resources/META-INF/native-image -jar build/libs/$JARFILE_NAME
```

The command we use for building with spring:
```
./gradlew nativeCompile distZip
```

**Comment by sendev1** [comment]
Any update on this issue.  I am running into same issue with GraalVM 22.1.0.

`
Fatal error: com.oracle.svm.core.util.VMError$HostedError: com.oracle.svm.core.util.VMError$HostedError: Registering type as reachable after analysis: AnalysisType<com.fasterxml.jackson.databind.JsonSerializer[], allocated: false, inHeap: false, reachable: false>
        at com.oracle.svm.core.util.VMError.shouldNotReachHere(VMError.java:72)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:678)
        at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:515)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:407)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:585)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:615)
Caused by: com.oracle.svm.core.util.VMError$HostedError: Registering type as reachable after analysis: AnalysisType<com.fasterxml.jackson.databind.JsonSerializer[], allocated: false, inHeap: false, reachable: false>
        at com.oracle.svm.core.util.VMError.shouldNotReachHere(VMError.java:68)
        at com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.registerAsReachable(AnalysisConstantReflectionProvider.java:332)
        at com.oracle.svm.hosted.ameta.HostedDynamicHubFeature.replace(HostedDynamicHubFeature.java:53)
        at com.oracle.graal.pointsto.meta.AnalysisUniverse.replaceObject(AnalysisUniverse.java:582)
        at com.oracle.svm.hosted.image.NativeImageHeap.addArrayElements(NativeImageHeap.java:576)
        at com.oracle.svm.hosted.image.NativeImageHeap.addObjectToImageHeap(NativeImageHeap.java:473)
        at com.oracle.svm.hosted.image.NativeImageHeap.addObject(NativeImageHeap.java:295)
        at com.oracle.svm.hosted.image.NativeImageHeap.processAddObjectWorklist(NativeImageHeap.java:598)
        at com.oracle.svm.hosted.image.NativeImageHeap.addTrailingObjects(NativeImageHeap.java:198)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:660)
        ... 5 more
Error: Image build request failed with exit status 1
`


I used @Twometer python script but still no luck.

I tried using this command
```
$JAVA_HOME/bin/java -DspringAot=true -agentlib:native-image-agent=config-merge-dir=src/main/resources/META-INF/native-image -jar build/libs/$JARFILE_NAME
````

and I get `ApplicationContext could not start as 'org.springframework.aot.StaticSpringFactories' that is generated by the Spring AOT plugin could not be found.`

@Twometer could you please help

**Comment by Twometer** [comment]
@sendev1 I can try taking a look at the file

**Comment by sendev1** [comment]
@Twometer I tried your reproducer but I am not getting any error. I just ran `./gradlew nativeRun` and native image compiled successfully and it ran 
<img width="1726" alt="image" src="https://user-images.githubusercontent.com/68568448/175904614-08c18787-940c-469f-ad40-d6f4ee6df2e4.png">

**Comment by sendev1** [comment]
well, I ran `regen-agent-configs.sh` script from reproducer and I tried to compile native image with that generated config and I am getting the error`com.oracle.svm.core.util.VMError$HostedError: Registering type as reachable after analysis: AnalysisType<org.springframework.boot.web.embedded.tomcat.TomcatProtocolHandlerCustomizer[], allocated: false, inHeap: false, reachable: false>
`
<img width="1737" alt="image" src="https://user-images.githubusercontent.com/68568448/175907800-19811e3d-3893-41eb-997d-7e5365772d75.png">

When I compared the checked in config and generated config by the script I see quite a bit of differences but the main deference are

1. `queryAllDeclaredMethods` in generated config is changed to `allDeclaredMethods` in the checked in config.
2. `queriedMethods`  in generated config is changed to `methods` in the checked in config.

I do not see `TomcatProtocolHandlerCustomizer` used anywhere in the checked in script. Did you make the `queryAllDeclaredMethods` and `queriedMethods` manually in the checked script. Please help me to understand how come they are `allDeclaredMethods` and `methods`.
left side -> checked config from github
rightside -> generated config using `regen-agent-configs.sh`
<img width="1730" alt="image" src="https://user-images.githubusercontent.com/68568448/175909083-aa36d5c9-98ac-49c0-b966-eaf36459bf9c.png">

**Comment by matheusmahl** [comment]
I was having the same issue with the agent generated config, even with the additional configs generated by the python script. But I've solved that by including this config in the reflect-config.json

{
  "name":"org.springframework.beans.factory.ObjectProvider",
  "allDeclaredFields":true,
  "queryAllDeclaredMethods":true,
  "queryAllPublicMethods":true
}

and changing the "queryAllDeclaredMethods" and "queryAllPublicMethods", to this: 

{
  "name":"org.springframework.boot.web.embedded.tomcat.TomcatServletWebServerFactory",
  "allDeclaredFields":true,
  "allDeclaredMethods":true,
  "allPublicMethods":true
},  

The python script didn't show this ObjectProvider config either.

I've found that class at the parameterTypes in this config that had the "tomcatServletWebServerFactory":

{
  "name":"org.springframework.boot.autoconfigure.web.servlet.ServletWebServerFactoryConfiguration$EmbeddedTomcat",
  "allDeclaredFields":true,
  "queryAllDeclaredMethods":true,
  "queryAllPublicMethods":true,
  "queryAllDeclaredConstructors":true,
  "methods":[
    {"name":"<init>","parameterTypes":[] }, 
    {"name":"tomcatServletWebServerFactory","parameterTypes":["org.springframework.beans.factory.ObjectProvider","org.springframework.beans.factory.ObjectProvider","org.springframework.beans.factory.ObjectProvider"] }
  ]
}

I hope this helps understand the issue.

**Comment by ksnovich** [comment]
I have the same problem. 
I had reflect-config.json generated by Spring Native, after that  I ran agent like this "$JAVA_HOME/bin/java -DspringAot=true -agentlib:native-image-agent=config-merge-dir ...." which merged old reflect-config.json with agent's data.
Also I tried to use [Twometer](https://github.com/Twometer) python script
Nothing helped....

But when I change "graalvm-ce-java11-22.1.0" to "graalvm-ce-java11-22.2.0-**dev**" the problem is gone.

Of course it isn't released yet.
Let's wait for the realease!

**Comment by TDtianzhenjiu** [comment]
hello @ksnovich 
22.2 already released
https://github.com/graalvm/graalvm-ce-builds/releases/tag/vm-22.2.0

But I still encounter this issue
have you resolved this issue by using graalvm-ce-java11-22.2.0-dev?

**Comment by wirthi** [comment]
I am closing this older ticket. This is based on Spring native, which has since been vastly improved and shipped as a mainline Spring feature. Can you please try with the latest version of Spring (Spring Boot 3 or newer), and reopen if the problem persists then?
